Framework/Catalog/src/ONCat.cpp

```diff
@@ -0,0 +1,449 @@
+ */
+void ONCat::refreshTokenIfNeeded() {
+  refreshTokenIfNeeded(DateAndTime::getCurrentTime());
+}
+
+/**
+ * See overloaded method.
+ *
+ * @param currentTime : Used in testing to specify a different time.
+ *
+ * @exception Mantid::Catalog::Exception::InvalidRefreshTokenError :
+ *   Thrown when the provider decides the current token cannot be refreshed.
+ */
+void ONCat::refreshTokenIfNeeded(const DateAndTime &currentTime) {
+  const auto currentToken = m_tokenStore->getToken();
+
+  if (m_flow == OAuthFlow::CLIENT_CREDENTIALS) {
+    if (currentToken && !currentToken->isExpired(currentTime)) {
+      return;
+    }
+
+    Poco::Net::HTMLForm form(Poco::Net::HTMLForm::ENCODING_MULTIPART);
+    form.set("client_id", m_clientId);
+    if (m_clientSecret) {
+      form.set("client_secret", m_clientSecret.get());
+    }
+    form.set("grant_type", "client_credentials");
+
+    m_internetHelper->reset();
+    m_internetHelper->setBody(form);
+
+    try {
+      std::stringstream ss;
+
+      const int statusCode =
+          m_internetHelper->sendRequest(m_url + "/oauth/token", ss);
+
+      if (statusCode == HTTPResponse::HTTP_OK) {
+        m_tokenStore->setToken(OAuthToken::fromJSONStream(ss));
+      }
+      g_log.debug() << "Token successfully refreshed." << std::endl;
+    } catch (InternetError &ie) {
+      throw CatalogError(ie.what());
+    }
+  } else if (m_flow == OAuthFlow::RESOURCE_OWNER_CREDENTIALS) {
+    if (!currentToken.is_initialized()) {
+      return;
+    }
+    if (!currentToken->isExpired(currentTime)) {
+      return;
+    }
+    const auto currentRefreshToken = currentToken->refreshToken();
+    if (!currentRefreshToken) {
+      return;
+    }
+
+    Poco::Net::HTMLForm form(Poco::Net::HTMLForm::ENCODING_MULTIPART);
+    form.set("client_id", m_clientId);
+    if (m_clientSecret) {
+      form.set("client_secret", m_clientSecret.get());
+    }
+    form.set("grant_type", "refresh_token");
+    form.set("refresh_token", currentRefreshToken.get());
+
+    m_internetHelper->reset();
+    m_internetHelper->setBody(form);
+
+    try {
+      std::stringstream ss;
+
+      const int statusCode =
+          m_internetHelper->sendRequest(m_url + "/oauth/token", ss);
+
+      if (statusCode == HTTPResponse::HTTP_OK) {
+        m_tokenStore->setToken(OAuthToken::fromJSONStream(ss));
+      }
+      g_log.debug() << "Token successfully refreshed." << std::endl;
+    } catch (InternetError &ie) {
+      if (ie.errorCode() == HTTPResponse::HTTP_UNAUTHORIZED) {
+        // As per OAuth spec, when a refresh token is no longer valid, we
+        // can consider ourselves logged out.
+        logout();
+        throw InvalidRefreshTokenError(
+            "You have been logged out.  Please login again.");
+      }
+      throw CatalogError(ie.what());
+    }
+  }
+}
+
+void ONCat::setInternetHelper(
+    std::shared_ptr<Mantid::Kernel::InternetHelper> internetHelper) {
+  m_internetHelper = std::move(internetHelper);
```

Comment:
Another reason why it looks like `std::unique_ptr` is better for `InternetHelper`.

Reply:
God knows what I was trying to do here.  I probably changed from an initial `unique_ptr` approach to using a `shared_ptr`, and just forgot to get rid of `std::move`, which wouldn't be required here if I just changed the method signature to use a reference:

```c++
void ONCat::setInternetHelper(const std::shared_ptr<Mantid::Kernel::InternetHelper> & internetHelper)
```